web/src/components/pages/dashboard/challenge/achievement.ts

```diff
@@ -0,0 +1,69 @@
+export type AchievementType =
+  | {
+      type: 'invite';
+    }
+  | {
+      type: 'contribute';
+      firstContribute: boolean;
+      firstStreak: boolean;
+      hasAchieved: boolean;
+    };
+export function handleAchievements(
+  achievement: AchievementType,
+  addAchievement: any,
+  setInviteContributeAchievement: any
+): void {
+  if (achievement.type === 'invite') {
+    if (JSON.parse(sessionStorage.getItem('firstInvite'))) {
+      addAchievement(50, 'You sent your first invite!');
+    }
+    if (
+      !JSON.parse(sessionStorage.getItem('hasAchieved')) &&
+      JSON.parse(sessionStorage.getItem('hasContributed'))
+    ) {
+      addAchievement(
+        50,
+        "You're on a roll! You sent an invite and contributed in the same session."
+      );
+      // Tell back-end user get unexpected achievement: invite + contribute in the same session
+      // Each user can only get once.
+      setInviteContributeAchievement();
+      // Remove items from session storage.
+      sessionStorage.removeItem('hasAchieved');
```

Comment:
As I mentioned above, I think our use of `sessionStorage` has leaked beyond its original intent. I would expect for this function to look something like:

```python
(pseudocode)

# User clicks the invite button. Response comes back from server.
if response.firstInvite:
    triggerAchievement('firstInvite')

if not response.hasEarnedInviteContributeSameSession:
    triggerAchievement('inviteContributeSameSession', { hasInvited: true })
    # Vice versa for contributions, which can use the same function.

# In triggerAchievement(type, options):
if type === 'inviteContributeSameSession':
    if (options.hasInvited):
        sessionStorage.setItem('hasInvited', true)
    if (options.hasContributed):
        sessionStorage.setItem('hasContributed', true)

    if sessionStorage.getItem('hasInvited') and sessionStorage.getItem('hasContributed'):
        # Handle adding the achievement.
        sessionStorage.removeItem('hasInvited')
        sessionStorage.removeItem('hasContributed')
        # From now on, the server's response.hasEarnedInviteContributeSameSession
        # will be true, so we won't get here… I think? These are rough thoughts so
        # let me know what you think.
```